File: Game_Of_Life_Python/Organisms/Plants/sosnowsky_hogweed.py
from Organisms.organism import OrganismInitialData
from Organisms.plant import Plant
from Organisms.animal import CollisionTypes


class SosnowskyHogweed(Plant):

    static_counter = 0

    def __init__(self, name, row, column, given_strength=-1, given_initiative=-1):
        strength = OrganismInitialData[name]["strength"]
        initiative = OrganismInitialData[name]["initiative"]
        character = OrganismInitialData[name]["character"]
        if given_strength != -1:
            strength = given_strength
        if given_initiative != -1:
            initiative = given_initiative
        super().__init__(strength, initiative, name,
                         character, row, column, "sosnowsky_hogweed.png")
        SosnowskyHogweed.static_counter += 1
        self.type = SosnowskyHogweed

    def can_kill(self, grid_board, row, column):
        return (grid_board[row][column] == 'w'
                or grid_board[row][column] == 's'
                or grid_board[row][column] == 'f'
                or grid_board[row][column] == 't'
                or grid_board[row][column] == 'a'
                or grid_board[row][column] == 'H')

    # Kill avery animal except cyber_sheep in it's neighbourhood
    def action(self, grid_board):
        self.age += 1
        rows_columns_to_remove = []
        for i in range(-1, 2):
            for j in range(-1, 2):

                if i == 0 and j == 0:
                    continue
                r, c = self.row + i, self.column + j

                if r >= 0 and r < len(grid_board) and c >= 0 and c < len(grid_board[r]) and self.can_kill(grid_board, r, c):
                    rows_columns_to_remove.append((r, c))
                    self.print_to_journal(f"Sosnowsky kill {grid_board[r][c]} at ({r}, {c})\n")
                    grid_board[r][c] = 'e'

        if len(rows_columns_to_remove):
            return rows_columns_to_remove
        self.print_to_journal(
            f"Sosnowsky at ({self.row}, {self.column}) will not kill\n")
        return None

    def collision(self, grid_board, organisms, current_index):
        
        for organism_index, organism in enumerate(organisms):

            if organism.get_position_row() == self.row and organism.get_position_column() == self.column and organism_index != current_index and organism.get_name() != "cyber_sheep":
                
                self.print_to_journal(f"{self.character}vs{organism.get_character()} at {self.row},{self.column} -> {organism.get_character()} eat {self.character}, {self.character} eat {organism.get_character()}\n")
                return CollisionTypes("Fight"), organism_index

        return self.default_plant_collision()


    def get_static_counter(self):
        return SosnowskyHogweed.static_counter

    def increase_static_counter(self):
        SosnowskyHogweed.static_counter += 1

    def decrease_static_counter(self):
        SosnowskyHogweed.static_counter -= 1

File: GameOfLife_Cpp/source/src/sosnowsky_hogweed.cpp
#include "sosnowsky_hogweed.hpp"
#include <cassert>
#include <iostream>

int SosnowskyHogweed::SOSNOWSKY_HOGWEED_STATIC_COUNTER = 0;

SosnowskyHogweed::SosnowskyHogweed(int row, int column)
    : Plant(10, 0, 0, "SosnowskyHogweed", 'O', row, column, "sosnowsky_hogweed.png", OrganismType::SOSNOWSKY_HOGWEED)
{
    SOSNOWSKY_HOGWEED_STATIC_COUNTER++;
}

SosnowskyHogweed::SosnowskyHogweed(int row, int column, int strength, int initiative, int age)
    : Plant(strength, initiative, age, "SosnowskyHogweed", 'O', row, column, "sosnowsky_hogweed.png", OrganismType::SOSNOWSKY_HOGWEED)
{
    SOSNOWSKY_HOGWEED_STATIC_COUNTER++;
}

SosnowskyHogweed::~SosnowskyHogweed()
{
    SOSNOWSKY_HOGWEED_STATIC_COUNTER--;
}

ActionResult SosnowskyHogweed::action(std::vector<std::vector<char>>& grid_board)
{
    std::vector<Point> points;
    for (int i = -1; i <= 1; i++)
    {
        for (int j = -1; j <= 1; j++)
        {
            if (i == 0 && j == 0)
                continue;
            int new_row = row + i;
            int new_column = column + j;
            if (!(new_row >= 0 && new_row < BOARD_HEIGHT && new_column >= 0 && new_column < BOARD_WIDTH))
                continue;
            assert(new_row >= 0 && new_row < BOARD_HEIGHT && new_column >= 0 && new_column < BOARD_WIDTH);
            if (can_eat(grid_board[new_row][new_column]))
            {
                points.push_back(Point(new_row, new_column));
                std::cout << this->get_name() << " at (" << row << ", " << column << ") eats " << grid_board[new_row][new_column] << " at (" << new_row << ", " << new_column << ")\n";
            }
        }
    }
    return ActionResult(ActionType::KILLING, points);
}

CollisionResult SosnowskyHogweed::collision(std::vector<std::vector<char>>& grid_board, std::vector<Organism*>& organisms, int current_index)
{
    return this->default_collision_plant();
}

int SosnowskyHogweed::get_static_counter()
{
    return SOSNOWSKY_HOGWEED_STATIC_COUNTER;
}

bool SosnowskyHogweed::can_eat(char ch) const
{
    return ch == 'w' || ch == 's' || ch == 'H' || ch == 'f' || ch == 'a' || ch == 't';
}